src/frequenz/sdk/timeseries/battery_pool/_metric_calculator.py

```diff
@@ -0,0 +1,478 @@
+# License: MIT
+# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
+
+"""Module that defines how to aggregate metrics from battery-inverter components."""
+import logging
+from abc import ABC, abstractmethod
+from datetime import datetime, timezone
+from typing import Dict, Generic, Iterable, List, Optional, Set, TypeVar
+
+from ... import microgrid
+from ...microgrid.component import ComponentCategory, ComponentMetricId
+from ...timeseries import Sample
+from ._component_metrics import ComponentMetricsData
+from ._result_types import Bound, CapacityMetrics, PowerBoundsMetrics, SoCMetrics
+
+_logger = logging.getLogger(__name__)
+_MIN_TIMESTAMP = datetime.min.replace(tzinfo=timezone.utc)
+
+
+def battery_inverter_mapping(batteries: Iterable[int]) -> Dict[int, int]:
+    """Create mapping between battery and adjacent inverter.
+
+    Args:
+        batteries: Set of batteries
+
+    Returns:
+        Mapping between battery and adjacent inverter.
+    """
+    graph = microgrid.get().component_graph
+    bat_inv_map: Dict[int, int] = {}
+    for battery_id in batteries:
+        try:
+            predecessors = graph.predecessors(battery_id)
+        except KeyError as err:
+            _logger.error(str(err))
+            continue
+
+        inverter_id = next(
+            (
+                comp.component_id
+                for comp in predecessors
+                if comp.category == ComponentCategory.INVERTER
+            ),
+            None,
+        )
+        if inverter_id is None:
+            _logger.warning("Battery %d has no adjacent inverter.", battery_id)
+        else:
+            bat_inv_map[battery_id] = inverter_id
+    return bat_inv_map
+
+
+# Formula output types class have no common interface
+# Print all possible types here.
+T = TypeVar("T", Sample, SoCMetrics, CapacityMetrics, PowerBoundsMetrics)
+
+
+class MetricCalculator(ABC, Generic[T]):
+    """Define how to calculate high level metrics from many components data.
+
+    It specifies:
+        * what components and metrics its needs to calculate the results,
+        * how to calculate the result,
+    """
+
+    def __init__(self, batteries: Set[int]) -> None:
+        """Create class instance.
+
+        Args:
+            batteries: From what batteries the data should be aggregated.
+        """
+        self._batteries = batteries
+
+    @classmethod
+    @abstractmethod
+    def name(cls) -> str:
+        """Return name of the formula.
+
+        Returns:
+            Name of the formula
+        """
+
+    @property
+    def batteries(self) -> Set[int]:
```

Comment:
Warning: This is using `typing.Set` which is analogous to `builtins.set`, which means the user will be able to mutate the set, which is probably not what you want since you made it a property. If you use [`collections.abc.Set`](https://docs.python.org/3.11/library/collections.abc.html#collections.abc.Set) instead, then is an interface for a immutable set, so the user won't be allowed to change it (if they run `mypy` that is, of course at runtime they can do whatever they want :D)

Reply:
Done, thanks!

Reply:
Unfortunately it doesn't work for python3.8 :( 
```
_____ ERROR collecting tests/timeseries/_battery_pool/test_battery_pool.py _____
tests/timeseries/_battery_pool/test_battery_pool.py:22: in <module>
    from frequenz.sdk.timeseries.battery_pool import (
src/frequenz/sdk/timeseries/battery_pool/__init__.py:7: in <module>
    from .battery_pool import BatteryPool
src/frequenz/sdk/timeseries/battery_pool/battery_pool.py:23: in <module>
    class BatteryPool:
src/frequenz/sdk/timeseries/battery_pool/battery_pool.py:40: in BatteryPool
    batteries_id: Set[int] | None = None,
E   TypeError: 'ABCMeta' object is not subscriptable
```

Reply:
I can surround it with ditto mark ( batteries_id: "Set[int]" | None = None, )
But doing it everywhere seems to much.
We can change it when we delete support for python3.8.